### PR TITLE
ref(alerts): Update rule status on dynamic alert rule update

### PR DESCRIPTION
**Pull request by ceorourke**
A classic "various fixes" including:

If we’re updating the alert rule status from dynamic to dynamic, we also need to update the status to be PENDING if the updated alert rule has enough data.

This also fixes setting the resolution to be equal to the time window for dynamic alerts which was initially added in https://github.com/getsentry/sentry/pull/77364 but missed a spot further down in the code, and set the timedelta to minutes when it should be seconds.

I noticed that we weren't updating the snuba query data that we pass to Seer if the aggregate, dataset, or query changes when the rule is updated, so I fixed that as well.

Closes https://getsentry.atlassian.net/browse/ALRT-281

### CODE REVIEW

File: src/sentry/incidents/logic.py [hunk 1]
```diff
@@ -936,22 +936,37 @@ def update_alert_rule(
...
                     "Your organization does not have access to this feature."
                 )
-
             if updated_fields.get("detection_type") == AlertRuleDetectionType.DYNAMIC and (
                 alert_rule.detection_type != AlertRuleDetectionType.DYNAMIC or query or aggregate
             ):
                 for k, v in updated_fields.items():
                     setattr(alert_rule, k, v)
 
+                for k, v in updated_query_fields.items():
```
Comment (ceorourke) [inline src/sentry/incidents/logic.py:945]: any snuba query updates were previously not getting passed to the Seer code until this was added

File: src/sentry/incidents/logic.py [hunk 2]
```diff
@@ -936,22 +936,37 @@ def update_alert_rule(
...
+                    elif k == "event_types":
+                        continue
+                    setattr(alert_rule.snuba_query, k, v)
+
                 try:
                     # NOTE: if adding a new metric alert type, take care to check that it's handled here
                     rule_status = send_historical_data_to_seer(
                         alert_rule=alert_rule,
                         project=projects[0] if projects else alert_rule.projects.get(),
+                        snuba_query=alert_rule.snuba_query,
```
Comment (ceorourke) [inline src/sentry/incidents/logic.py:959]: since we fetch the snuba query from the db in `send_historical_data_to_seer` it was missing the updates unless I passed it like this

File: src/sentry/incidents/logic.py [hunk 3]
```diff
@@ -936,22 +936,37 @@ def update_alert_rule(
...
+                        event_types=updated_query_fields.get("event_types"),
```
Comment (ceorourke) [inline src/sentry/incidents/logic.py:960]: `event_types` is in another table and updated slightly differently, (see [here](https://github.com/getsentry/sentry/blob/master/src/sentry/snuba/subscriptions.py#L120-L130)) so I had to pass it in directly for updates

File: src/sentry/incidents/logic.py [hunk 4]
```diff
@@ -994,7 +1009,17 @@ def update_alert_rule(
                 "time_window", timedelta(seconds=snuba_query.time_window)
             )
             updated_query_fields.setdefault("event_types", None)
-            updated_query_fields.setdefault("resolution", timedelta(seconds=snuba_query.resolution))
+            if detection_type and detection_type == AlertRuleDetectionType.DYNAMIC:
```
Comment (ceorourke) [inline src/sentry/incidents/logic.py:1012]: this is the piece that was missing from the [previous PR](https://github.com/getsentry/sentry/pull/77364) that set the resolution to the time window, it would be overwritten here

File: src/sentry/seer/anomaly_detection/store_data.py [hunk 5]
```diff
@@ -57,16 +58,28 @@ def _get_start_and_end_indices(data: list[TimeSeriesPoint]) -> tuple[int, int]:
...
     """
-    snuba_query = SnubaQuery.objects.get(id=alert_rule.snuba_query_id)
+    if not snuba_query:
+        snuba_query = SnubaQuery.objects.get(id=alert_rule.snuba_query_id)
     window_min = int(snuba_query.time_window / 60)
-    dataset = get_dataset(snuba_query.dataset)
+    dataset = get_dataset_from_label(snuba_query.dataset)
     query_columns = get_query_columns([snuba_query.aggregate], snuba_query.time_window)
+    if not event_types:
+        event_types = snuba_query.event_types
```
Comment (saponifi3d) [inline src/sentry/seer/anomaly_detection/store_data.py:76]: is `snuba_query.event_types` always set? (we may want to provide a default if this is required or throw an exception if it's a required field)

File: src/sentry/seer/anomaly_detection/utils.py [hunk 6]
```diff
@@ -160,13 +164,24 @@ def format_historical_data(
...
 
 
+def get_dataset_from_label(dataset_label: str):
+    if dataset_label == "events":
+        # DATASET_OPTIONS expects the name 'errors'
+        dataset_label = "errors"
+    elif dataset_label in ["generic_metrics", "transactions"]:
+        # XXX: performance alerts dataset differs locally vs in prod
+        dataset_label = "discover"
+    return get_dataset(dataset_label)
```
Comment (saponifi3d) [inline src/sentry/seer/anomaly_detection/utils.py:174]: is the default case that we want to pass through to here? will this throw if it's an unexpected dataset?

File: src/sentry/seer/anomaly_detection/utils.py [hunk 7]
```diff
@@ -44,18 +46,20 @@ def get_snuba_query_string(snuba_query: SnubaQuery) -> str:
...
-            SNUBA_QUERY_EVENT_TYPE_TO_STRING[event_type] for event_type in snuba_query.event_types
+            SNUBA_QUERY_EVENT_TYPE_TO_STRING[event_type] for event_type in event_types
         ]
         event_types_string = "(event.type:["
         event_types_string += ", ".join(event_types_list)
         event_types_string += "])"
     else:
         # e.g. (is:unresolved) AND (event.type:error)
-        snuba_query_event_type_string = SNUBA_QUERY_EVENT_TYPE_TO_STRING[snuba_query.event_types[0]]
+        snuba_query_event_type_string = SNUBA_QUERY_EVENT_TYPE_TO_STRING[event_types[0]]
```
Comment (saponifi3d) [inline src/sentry/seer/anomaly_detection/utils.py:62]: can we merge these cases? not sure the difference of the events_types_string if it's a list with 1 event_type vs the singular event type.

File: src/sentry/seer/anomaly_detection/utils.py [hunk 8]
```diff
@@ -44,18 +46,20 @@ def get_snuba_query_string(snuba_query: SnubaQuery) -> str:
         SnubaQueryEventType.EventType.DEFAULT: "default",
```
Comment (saponifi3d) [inline src/sentry/seer/anomaly_detection/utils.py:46]: not part of this change set, but might be nice to move this static variable out of the method scope so it's not redefined each method invocation.

File: tests/sentry/incidents/test_logic.py [hunk 9]
```diff
@@ -1751,6 +1751,119 @@ def test_update_alert_rule_static_to_dynamic_enough_data(self, mock_seer_request
         assert mock_seer_request.call_count == 1
         assert alert_rule.status == AlertRuleStatus.PENDING.value
 
+    @with_feature("organizations:anomaly-detection-alerts")
+    @patch(
+        "sentry.seer.anomaly_detection.store_data.seer_anomaly_detection_connection_pool.urlopen"
+    )
+    def test_update_dynamic_alert_not_enough_to_pending(self, mock_seer_request):
```
Comment (saponifi3d) [inline tests/sentry/incidents/test_logic.py:1758]: note to self - this is freezing time - it's just pretty hidden in this diff.

**Review by saponifi3d — COMMENTED** [review]
this is a bit hard to review because it's diving into a few places and has a few different goals.

---

i think that's a bit of a code smell though, for example the `update_alert_rule` method is pretty large, and this is adding more conditional logic to it. it'd be a bit easier to grok what the changes and updates for a dynamic alert are if there was a single control flow (if statement in this case). that would allow us to have all the "every alert does this stuff" and then after that code, have "if it's an anomaly detection type, also do more things". 

For this PR, we could just focus on decomposing the specifics for just the dynamic alerts. if the code had a little clearer structure to it, then it'd be a lot easier to understand the isolated changes that are required to seer as well.

File: src/sentry/incidents/logic.py (hunk 1 above)
Comment (ceorourke) [reply src/sentry/incidents/logic.py:945]: That check is saying "if the rule type is changing to become a dynamic type OR it's already dynamic and the update changes either the query or the aggregate..." then we need to refetch the historical data and send it to Seer because either it wasn't a dynamic type before (so Seer wouldn't have that data) or the snuba query has changed so the historical data has changed

File: src/sentry/seer/anomaly_detection/store_data.py (hunk 5 above)
Comment (ceorourke) [reply src/sentry/seer/anomaly_detection/store_data.py:76]: Yes a snuba query will always have event types on it, but I changed this function to accept both `snuba_query` and `event_types` because this is happening during an alert rule _update_ - to avoid a bad state, we don't want to save the changes until we've successfully sent the data to Seer, so I can't pull the snuba query off of the alert rule (it may have changed) and the event types on the snuba query get updated separately

File: src/sentry/seer/anomaly_detection/utils.py (hunk 7 above)
Comment (ceorourke) [reply src/sentry/seer/anomaly_detection/utils.py:62]: The if and the else case? I left comments on what the strings look like. The only time it'll have more than one is if you pick both from the dropdown here: 
Screenshot 2024-09-26 at 3 26 57 PM

File: src/sentry/seer/anomaly_detection/utils.py (hunk 6 above)
Comment (ceorourke) [reply src/sentry/seer/anomaly_detection/utils.py:174]: If it's unexpected something is seriously wrong with how we're creating alert rules, but I can throw something to be safe

File: src/sentry/seer/anomaly_detection/utils.py (hunk 8 above)
Comment (ceorourke) [reply src/sentry/seer/anomaly_detection/utils.py]: I moved it out 👍
